caffe2/core/dispatch/DispatchTable.h

```diff
@@ -0,0 +1,150 @@
+#pragma once
+
+#include "caffe2/utils/flat_hash_map/flat_hash_map.h"
+#include "caffe2/utils/Metaprogramming.h"
+#include "caffe2/core/dispatch/OpSchema.h"
+
+#include <type_traits>
+#include <array>
+#include <unordered_map>
+#include <iostream>
+#include <mutex>
+
+namespace c10 {
+
+namespace details {
+
+/// Internal implementation of the operator as a thread-safe hash table.
+template<class Key>
+class ThreadsafeOperatorTable_ final {
+public:
+    // TODO The current implementation below does not have the correct correctness characteristics
+    // which we need.  It's worth spelling out exactly what we need:
+    //
+    //  - We need LOCK FREE read access to the table (as per the performance benchmark
```

Comment:
probably creating a different map and doing lazy swap with freeing the previous map later is the way to go

Reply:
That sounds like a good idea. I added is as one of the options to the TODO.